Feature/ah/cgo 1127 UI landing page

arthur-hermann opened the pull request:
Recipe images were hardcoded, causing the landing page to show only one image. This fixes the handling of recipe images, enabling recipe images to be correctly retrieved and displayed to users. Displaying recipe images in the landing page will enable users to press on a recipe's image to see the detailed view for that recipe.  

API: https://spoonacular.com/food-api/docs

arthur-hermann committed:
Implements search functionality for recipes.

arthur-hermann committed:
Added Search Bar and UI Collection View

arthur-hermann committed:
removed duplicate files

arthur-hermann committed:
made changes to viewModel and test classes.

arthur-hermann committed:
- Made changes to recipeListViewModel
- Commented out unit tests
- Made a recipe's image optional

arthur-hermann committed:
- Removed testing files
- Made changes to RecipeListViewModel

arthur-hermann committed:
- moved private methods to an extension at the end of the file
- moved RecipeError enum to seperate file

arthur-hermann committed:
- Added image display and search bar

arthur-hermann committed:
removed file that shouldn't be tracked.

arthur-hermann committed:
- fixed indentation issues in RecipeListViewModel

arthur-hermann committed:
Merge branch 'feature/ah/CGO-1129_Network-Landing-Page'

#Conflicts:
#	Supporting/Assets.xcassets/AccentColor.colorset/Contents.json
#	Supporting/Assets.xcassets/AppIcon.appiconset/Contents.json
#	Supporting/Assets.xcassets/Contents.json
#	Supporting/Info.plist
#	recipe-app.xcodeproj/xcuserdata/arthur.hermann.xcuserdatad/xcdebugger/Breakpoints_v2.xcbkptlist
#	recipe-app/Assets.xcassets/AccentColor.colorset/Contents.json
#	recipe-app/Assets.xcassets/AppIcon.appiconset/Contents.json
#	recipe-app/Assets.xcassets/Contents.json
#	recipe-app/Supporting/AppDelegate.swift
#	recipe-app/View/RecipeListView.swift
#	recipe-app/ViewController.swift

arthur-hermann committed:
- added fetchImage method

arthur-hermann committed:
- Images now work in UI

arthur-hermann committed:
Delete Supporting directory

deleting duplicate files

arthur-hermann committed:
Delete View directory

deleting duplicate files

arthur-hermann committed:
Delete Assets.xcassets directory

deleting duplicate files

arthur-hermann committed:
Delete Info.plist

deleting duplicate files

ilayda-Kodall commented on recipe-app/Supporting/AppDelegate.swift:
```diff
@@ -10,8 +10,6 @@ import UIKit
 @main
 class AppDelegate: UIResponder, UIApplicationDelegate {
 
-
```
Let's remove this if you are not using it.

ilayda-Kodall commented on recipe-app/Supporting/Assets.xcassets/germany.imageset/Contents.json:
```diff
@@ -0,0 +1,23 @@
+{
```
Why do you need these?

ilayda-Kodall commented on recipe-app/View/RecipeListViewController.swift:
```diff
@@ -0,0 +1,112 @@
+//
+//  ViewController.swift
```
Let's make sure this matches the file name.

ilayda-Kodall commented on recipe-app/Supporting/SceneDelegate.swift:
```diff
@@ -13,10 +13,11 @@ class SceneDelegate: UIResponder, UIWindowSceneDelegate {
-        guard let _ = (scene as? UIWindowScene) else { return }
+        guard let windowScene = (scene as? UIWindowScene) else {return}
+        let window = UIWindow(windowScene: windowScene)
+        window.rootViewController = ViewController()
```
I don't think your root view controller is named ViewController anymore.

ilayda-Kodall commented on recipe-app/View/CustomCollectionViewCell.swift:
```diff
@@ -0,0 +1,49 @@
+    
+    static let identifier = "CustomCollectionViewCell"
+    
+    let myImageView: UIImageView = {
```
Private

ilayda-Kodall commented on recipe-app/View/CustomCollectionViewCell.swift:
```diff
@@ -0,0 +1,49 @@
+
+final class CustomCollectionViewCell: UICollectionViewCell {
+    
+    static let identifier = "CustomCollectionViewCell"
```
Optional: We don't need this dependency. You are using it on one place so its fine to use a sting.

ilayda-Kodall commented on recipe-app/View/CustomCollectionViewCell.swift:
```diff
@@ -0,0 +1,49 @@
+        self.addSubview(myImageView)
+        myImageView.translatesAutoresizingMaskIntoConstraints = false
+        
+        NSLayoutConstraint.activate([
```
We can create a separate method for configuring constraint so it's clear what each method does. 
```suggestion
        private let myImageView: UIImageView = {
        let imageView = UIImageView()
        imageView.contentMode = .scaleAspectFill
        imageView.image = UIImage(systemName: "image")
        imageView.tintColor = .white
        imageView.clipsToBounds = true
        imageView.translatesAutoresizingMaskIntoConstraints = false
        return imageView
    }()
    
    private func setUp() {
        self.addSubview(myImageView)
        configureConstraints()
    }
    
    private func configureConstraints() {
        NSLayoutConstraint.activate([
            myImageView.topAnchor.constraint(equalTo: self.topAnchor),
            myImageView.bottomAnchor.constraint(equalTo: self.bottomAnchor),
            myImageView.trailingAnchor.constraint(equalTo: self.trailingAnchor),
            myImageView.leadingAnchor.constraint(equalTo: self.leadingAnchor)
        ])
    }
```

ilayda-Kodall commented on recipe-app/View/RecipeListViewController.swift:
```diff
@@ -0,0 +1,112 @@
+    
+    var recipeListViewModel = RecipeListViewModel()
+    
+    let searchBar: UISearchBar = {
```
private

ilayda-Kodall commented on recipe-app/View/RecipeListViewController.swift:
```diff
@@ -0,0 +1,112 @@
+        return searchBar
+    }()
+    
+    let collectionView: UICollectionView = {
```
private

ilayda-Kodall commented on recipe-app/View/RecipeListViewController.swift:
```diff
@@ -0,0 +1,112 @@
+        searchBar.translatesAutoresizingMaskIntoConstraints = false
+        collectionView.translatesAutoresizingMaskIntoConstraints = false
+        
+        NSLayoutConstraint.activate([
```
Let's move this to another method as well.

ilayda-Kodall commented on recipe-app/View/RecipeListViewController.swift:
```diff
@@ -0,0 +1,112 @@
+        self.view.addSubview(collectionView)
+        self.view.addSubview(searchBar)
+        
+        searchBar.translatesAutoresizingMaskIntoConstraints = false
```
You can set these in their corresponding components.
```suggestion
       private let collectionView: UICollectionView = {
        let layout = UICollectionViewFlowLayout()
        layout.scrollDirection = .vertical

        let collectionView = UICollectionView(frame: .zero, collectionViewLayout: layout)
        collectionView.register(CustomCollectionViewCell.self, forCellWithReuseIdentifier: "CustomCollectionViewCell")
        collectionView.translatesAutoresizingMaskIntoConstraints =  false
        return collectionView
    }()
```

ilayda-Kodall commented on recipe-app/ViewModel/RecipeListViewModel.swift:
```diff
@@ -0,0 +1,96 @@
+        return recipes
+    }
+    
+    func fetchImages(_ query: String, completion: @escaping () -> Void) {
```
Optional: You can make this a UIImageView extension for reusability.

```suggestion
    extension UIImageView {
    func setImage(from urlString: String, placeholder: UIImage? = nil) {
        guard let url = URL(string: "<YourImagePath>" + urlString) else {
            self.image = placeholder
            return
        }

        self.image = placeholder

        let task = URLSession.shared.dataTask(with: url) { [weak self] (data, response, error) in
            guard let imageData = data, let image = UIImage(data: imageData) else { return }

            DispatchQueue.main.async {
                self?.image = image
            }
        }

        task.resume()
    }
}
```

ilayda-Kodall commented on recipe-app/.gitignore:
```diff
@@ -0,0 +1 @@
+
```
🤔

ilayda-Kodall commented on recipe-app/Model/Recipe.swift:
```diff
@@ -0,0 +1,14 @@
+struct Recipe: Codable {
+    let id: Int
+    var title: String
+    var image: String?
```
Are you 100% sure? My comment was a question.

ilayda-Kodall commented on recipe-app/ViewModel/RecipeListViewModel.swift:
```diff
@@ -0,0 +1,96 @@
+            let jsonRecipes = try decoder.decode(Recipes.self, from: json)
+            self.recipes = jsonRecipes.results
+        }
+        catch {
```
We can just throw if you are doing nothing on the catch.

ilayda-Kodall commented on recipe-app/ViewModel/RecipeListViewModel.swift:
```diff
@@ -0,0 +1,96 @@
+    private func makeURL(query: String) -> URL? {
+        var components = URLComponents(string: "https://api.spoonacular.com/recipes/complexSearch")
+        components?.queryItems = [
+            URLQueryItem(name: "apiKey", value: "d71eab7547c442199b0231aefdc871a7"),
```
Please research how you can store this key securely.

ilayda-Kodall commented on recipe-app/ViewModel/RecipeListViewModel.swift:
```diff
@@ -0,0 +1,96 @@
+             }
+           }.resume()
+         }
+       }
```
Please check indentation of your files.

ilayda-Kodall commented on recipe-app/View/CustomCollectionViewCell.swift:
```diff
@@ -0,0 +1,49 @@
+    }
+}
+
+
```
extra space line

ilayda-Kodall commented on recipe-app.xcodeproj/xcuserdata/arthur.hermann.xcuserdatad/xcdebugger/Breakpoints_v2.xcbkptlist:
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This shouldn't be on your PR. Let's create a .gitignore so this stops happening.

arthur-hermann committed:
- made nameClean optional because some recipes do not have it

- made changes to constraints to remove unecessary white space

# Conflicts:
#	recipe-app/Model/Ingredient.swift
#	recipe-app/View/RecipeDetailViewController.swift
#	recipe-app/View/RecipeListViewController.swift
#	recipe-app/ViewModel/RecipeDetailViewModel.swift
#	recipe-app/ViewModel/RecipeListViewModel.swift

arthur-hermann committed:
- fixed indentation issues

# Conflicts:
#	recipe-app/View/RecipeDetailViewController.swift
#	recipe-app/ViewModel/RecipeListViewModel.swift

arthur-hermann committed:
Merge remote-tracking branch 'refs/remotes/origin/feature/ah/CGO-1127_UI-Landing-Page'

arthur-hermann committed:
Delete recipe-app/Supporting/Assets.xcassets directory
